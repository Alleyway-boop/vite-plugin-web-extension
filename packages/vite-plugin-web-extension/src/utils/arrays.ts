export function compact<T>(array: Array<T | undefined>): T[] {
  return array.filter((item) => item != null) as T[];
}
