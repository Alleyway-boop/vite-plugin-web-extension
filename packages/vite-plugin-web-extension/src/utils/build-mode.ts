export enum BuildMode {
  BUILD,
  WATCH,
  DEV,
}
