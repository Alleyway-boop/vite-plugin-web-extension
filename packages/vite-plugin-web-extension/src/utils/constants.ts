export const MANIFEST_LOADER_PLUGIN_NAME = `web-extension:manifest`;
export const LABELED_STEP_PLUGIN_NAME = `web-extension:labeled-step`;
