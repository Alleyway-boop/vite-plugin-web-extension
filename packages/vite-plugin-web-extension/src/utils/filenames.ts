export function entryFilenameToOutput(filename: string): string {
  return filename
    .replace(/.(tsx?|jsx?|json)$/, ".js")
    .replace(/.(scss|sass|less|stylus)$/, ".css");
}

export function entryFilenameToInput(filename: string): string {
  return filename.substring(0, filename.lastIndexOf("."));
}
