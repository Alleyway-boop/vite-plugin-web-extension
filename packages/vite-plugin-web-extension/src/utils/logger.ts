import { PLUGIN_NAME, VERBOSE_LOGGER } from "./constants";

export interface Logger {
  verbose(message: string): void;
  log(message: string): void;
  warn(message: string): void;
  error(message: string, error: unknown): void;
}

const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM = "\x1b[2m";
const RED = "\x1b[91m";
const GREEN = "\x1b[92m";
const YELLOW = "\x1b[93m";

export function createLogger(verbose?: boolean): Logger {
  return {
    verbose(message: string) {
      if (!verbose) return;
      console.debug(
        message
          .split("\n")
          .map(
            (line) => `  ${BOLD}${DIM}vite:${VERBOSE_LOGGER}${RESET} ${line}`
          )
          .join("\n")
      );
    },
    log(message: string) {
      console.log(
        message
          .split("\n")
          .map((line) => `${BOLD}${GREEN}[${PLUGIN_NAME}]${RESET} ${line}`)
          .join("\n")
      );
    },
    warn(message: string) {
      console.warn(
        message
          .split("\n")
          .map(
            (line) => `${BOLD}${YELLOW}[${PLUGIN_NAME}] WARN: ${line}${RESET}`
          )
          .join("\n")
      );
    },
    error(message: string, err: unknown) {
      console.error(
        message
          .split("\n")
          .map((line) => `${BOLD}${RED}[${PLUGIN_NAME}] ERROR: ${line}${RESET}`)
          .join("\n")
      );
      console.error(err);
    },
  };
}
