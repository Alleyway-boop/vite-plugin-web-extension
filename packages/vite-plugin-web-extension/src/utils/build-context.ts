import { OutputAsset, OutputChunk, RollupOutput } from "rollup";
import { inspect } from "util";
import * as Vite from "vite";
import { Manifest } from "webextension-polyfill";
import { PluginOptions } from "../options";
import { labeledStepPlugin } from "../plugins/labeled-step-plugin";
import { compact } from "./arrays";
import { BuildMode } from "./build-mode";
import { PLUGIN_NAME } from "./constants";
import { entryFilenameToInput } from "./filenames";
import { Logger } from "./logger";
import { mergeConfigs } from "./merge-configs";
import path from "node:path";
import { getRootDir } from "./paths";
import uniqBy from "lodash.uniqby";

export interface BuildContext {
  /**
   * Based on the baseConfig and new manifest, rebuild all the entrypoints and update the bundle
   * map.
   */
  rebuild(
    baseConfig: Vite.InlineConfig,
    manifest: any,
    mode: BuildMode
  ): Promise<void>;
  getBundles(): Array<OutputChunk | OutputAsset>;
}

/**
 * Keeps track of everything that needs to be build for the extension, and orchastraits the actual
 * building of each part of the extension.
 */
export function createBuildContext({
  pluginOptions,
  logger,
}: {
  pluginOptions: PluginOptions;
  logger: Logger;
}): BuildContext {
  /**
   * Tracks an each input path relative to the Vite root, to their output filename and a list of
   * generated assets.
   */
  let bundles: Array<OutputChunk | OutputAsset> = [];

  //#region Build Config Generation
  async function generateBuildConfigs(
    baseConfig: Vite.InlineConfig,
    manifest: any
  ) {
    const entryConfigs = await generateBuildConfigsFromManifest(
      baseConfig,
      manifest
    );
    const totalEntries = entryConfigs.length;
    const finalConfigs = entryConfigs
      .map((entryConfig, i) =>
        mergeConfigs(baseConfig, entryConfig, {
          // We shouldn't clear the screen for these internal builds
          clearScreen: false,
          // Don't copy static assets for the lib builds - already done during manifest build
          publicDir: false,
          // Don't empty the outDir, this is handled in the parent, manifest build process
          build: { emptyOutDir: false },
          plugins: [
            // Print a message before starting each step
            labeledStepPlugin(logger, totalEntries, i),
          ],
          // logLevel: "warn",
        })
      )
      // Exclude this plugin from child builds to break recursion
      .map((config) => removePlugin(config, PLUGIN_NAME));
    return finalConfigs;
  }

  async function generateBuildConfigsFromManifest(
    baseConfig: Vite.InlineConfig,
    manifest: any
  ): Promise<Vite.InlineConfig[]> {
    const configs: Vite.InlineConfig[] = [];
    const alreadyIncluded: Record<string, boolean> = {};

    const addConfig = (config: Vite.InlineConfig) => configs.push(config);
    const addHtmlConfig = (entries: string[]) => {
      const htmlConfig = mergeConfigs(
        {
          build: {
            rollupOptions: {
              input: entries.reduce<Record<string, string>>((input, entry) => {
                input[entryFilenameToInput(entry)] = path.resolve(
                  getRootDir(baseConfig),
                  entry
                );
                return input;
              }, {}),
            },
          },
        },
        pluginOptions.htmlViteConfig ?? {}
      );
      addConfig(htmlConfig);
    };
    const addScriptConfig = (entry: string) => {
      if (alreadyIncluded[entry]) return;
      alreadyIncluded[entry] = true;

      const moduleId = entryFilenameToInput(entry);
      /**
       * "content-scripts/some-script/index" -> "content-scripts/some-script/"
       * "some-script" -> ""
       */
      const outputDir = moduleId.includes("/")
        ? path.dirname(moduleId) + "/"
        : "";
      const scriptConfig = mergeConfigs(
        {
          build: {
            rollupOptions: {
              input: {
                [moduleId]: path.resolve(getRootDir(baseConfig), entry),
              },
              output: {
                // Configure the output filenames so they appear in the same folder
                // - content-scripts/some-script/index.<hash>.js
                // - content-scripts/some-script/index.<hash>.css
                entryFileNames: `[name].[hash].js`,
                assetFileNames: `${outputDir}[name].[hash].[ext]`,
              },
            },
          },
        },
        pluginOptions.scriptViteConfig ?? {}
      );
      addConfig(scriptConfig);
    };

    const { additionalHtmlInputs, additionalScriptInputs } =
      separateAdditionalInputs(pluginOptions.additionalInputs);

    // HTML Pages - multi-page-mode
    const htmlEntries = createEntryList([
      manifest.action?.default_popup,
      manifest.devtools_page,
      manifest.options_page,
      manifest.options_ui?.page,
      manifest.browser_action?.default_popup,
      manifest.page_action?.default_popup,
      manifest.sidebar_action?.default_panel,
      additionalHtmlInputs,
    ]);
    if (htmlEntries.length > 0) addHtmlConfig(htmlEntries);

    // Sandbox - multi-page-mode
    const sandboxEntries = createEntryList([manifest.sandbox?.pages]);
    if (sandboxEntries.length > 0) addHtmlConfig(sandboxEntries);

    // Every Background Script gets it's own config
    const backgroundScripts = createEntryList([
      manifest.background?.service_worker,
      manifest.background?.page,
      manifest.background?.scripts,
    ]);
    backgroundScripts?.forEach(addScriptConfig);

    // Every Content Script gets it's own config
    manifest.content_scripts?.forEach((cs: Manifest.ContentScript) => {
      const csEntries = createEntryList([cs.js, cs.css]);
      csEntries.forEach(addScriptConfig);
    });

    // Each additional script gets it's own config
    additionalScriptInputs?.forEach(addScriptConfig);

    if (configs.length === 0)
      throw Error(
        "No inputs found in manifest.json. Set `options.verbose = true` for more details."
      );

    return configs;
  }

  function createEntryList(
    a: Array<string | string[] | undefined> | undefined
  ): string[] {
    return compact<string>(a?.flat() ?? []);
  }

  function separateAdditionalInputs(additionalInputs?: string[]) {
    const additionalScriptInputs: string[] = [];
    const additionalHtmlInputs: string[] = [];
    additionalInputs?.forEach((additionalInput) => {
      if (additionalInput.endsWith("html"))
        additionalHtmlInputs.push(additionalInput);
      else additionalScriptInputs.push(additionalInput);
    });
    return { additionalScriptInputs, additionalHtmlInputs };
  }

  function removePlugin(
    config: Vite.InlineConfig,
    pluginName: string
  ): Vite.InlineConfig {
    return {
      ...config,
      plugins:
        config.plugins?.filter(
          (plugin) =>
            plugin && (!("name" in plugin) || plugin.name !== pluginName)
        ) ?? [],
    };
  }
  //#endregion

  //#region Build Modes

  return {
    async rebuild(baseConfig, manifest, mode) {
      const buildConfigs = await generateBuildConfigs(baseConfig, manifest);
      // Print configs deep enough to include rollup inputs
      logger.verbose("Final configs: " + inspect(buildConfigs, undefined, 7));

      const newBundles: Array<OutputChunk | OutputAsset> = [];
      for (const config of buildConfigs) {
        const output = await Vite.build(config);
        if (Array.isArray(output)) {
          newBundles.push(...output.map((o) => o.output).flat());
        } else if ("output" in output) {
          newBundles.push(...output.output);
        } else {
          // In watch mode, wait until it's built once
          await new Promise<void>((res, rej) => {
            output.addListener("event", (e) => {
              switch (e.code) {
                case "BUNDLE_END":
                  res();
                  break;
                case "ERROR":
                  rej(e.error);
                  break;
              }
            });
          });
        }
      }
      bundles = uniqBy(newBundles, "fileName");
    },
    getBundles() {
      return bundles;
    },
  };
}
