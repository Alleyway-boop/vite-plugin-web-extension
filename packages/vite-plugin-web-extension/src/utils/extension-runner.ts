import { InlineConfig } from "vite";
import webExt from "web-ext";
import { PluginOptions } from "../options";
import { getOutDir, getRootDir } from "./paths";
import * as webExtLogger from "web-ext/util/logger";
import { cosmiconfig } from "cosmiconfig";
import { Logger } from "./logger";
import { inspect } from "util";

export interface ExtensionRunner {
  reload(): Promise<void>;
  exit(): Promise<void>;
}

export async function startWebExt(options: {
  pluginOptions: PluginOptions;
  config: InlineConfig;
  logger: Logger;
}): Promise<ExtensionRunner> {
  const { pluginOptions, config, logger } = options;

  //#region Config File
  const userConfig = await cosmiconfig("web-ext", {
    packageProp: "webExt",
    searchPlaces: [
      // Cosmiconfig Defaults
      "package.json",
      ".web-extrc",
      ".web-extrc.json",
      ".web-extrc.yaml",
      ".web-extrc.yml",
      ".web-extrc.js",
      ".web-extrc.cjs",
      "web-ext.config.js",
      "web-ext.config.cjs",
      // web-ext equivalents:
      // https://extensionworkshop.com/documentation/develop/getting-started-with-web-ext/#automatic-discovery-of-configuration-files
      "web-ext-config.js",
      "web-ext-config.cjs",
      ".web-ext-config.js",
      ".web-ext-config.cjs",
    ],
  }).search(getRootDir(config));
  if (userConfig != null) {
    logger.verbose(`web-ext config discovered at ${userConfig.filepath}:`);
    logger.verbose(inspect(userConfig.config));
  } else {
    logger.verbose(`No web-ext config discovered`);
  }
  //#endregion

  // Customize what is logged
  webExtLogger.consoleStream.write = ({ level, msg, name }) => {
    if (level >= ERROR_LOG_LEVEL) logger.error(name, msg);
    if (level >= WARN_LOG_LEVEL) logger.warn(msg);
  };
  const runner = await webExt.cmd.run(
    {
      // Open the browser passed in plugin options, falling back to a Chromium install
      target:
        pluginOptions.browser === null || pluginOptions.browser === "firefox"
          ? null
          : "chromium",
      ...userConfig?.config,
      // These options are required or the CLI freezes on linux
      sourceDir: getOutDir(config),
      noReload: true,
      noInput: true,
    },
    {
      // Don't call `process.exit(0)` after starting web-ext
      shouldExitProgram: false,
    }
  );
  return {
    reload: () => runner.reloadAllExtensions(),
    exit: () => runner.exit(),
  };
}

// https://github.com/mozilla/web-ext/blob/e37e60a2738478f512f1255c537133321f301771/src/util/logger.js#L12
const WARN_LOG_LEVEL = 40;
const ERROR_LOG_LEVEL = 50;
