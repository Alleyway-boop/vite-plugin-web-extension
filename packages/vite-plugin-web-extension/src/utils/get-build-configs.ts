import { ConfigEnv, mergeConfig, UserConfig } from "vite";
import { PluginOptions } from "../options";
import { Logger } from "./logger";
import { manifestWriter } from "../plugins/manifest-writer";
import { compact } from "./arrays";
import { entryFilenameToOutput } from "./filenames";
import { Manifest } from "webextension-polyfill";
import { InputOption } from "rollup";

const GENERATED_PREFIX = "generated:";

export interface ParseManifestOptions {
  baseConfig: UserConfig;
  env: ConfigEnv;
  pluginOptions: PluginOptions;
  logger: Logger;
}

export async function getBuildConfigs(options: {
  baseConfig: UserConfig;
  env: ConfigEnv;
  pluginOptions: PluginOptions;
  logger: Logger;
}): Promise<UserConfig[]> {
  const { pluginOptions, baseConfig } = options;
  const { plugin: manifestWriterPlugin, loadManifest } =
    manifestWriter(options);
  const manifest = await loadManifest();

  const configs: UserConfig[] = [];
  const addConfig = (newConfig: UserConfig) =>
    configs.push(mergeConfig(baseConfig, newConfig, true));

  const alreadyBuilt: Record<string, boolean> = {};
  const addLibModeConfig = (entry: string) => {
    if (alreadyBuilt[entry]) return;
    alreadyBuilt[entry] = true;

    const newConfig: UserConfig = {
      clearScreen: false,
      // Don't copy static assets for the lib builds - already done during multi-page builds
      publicDir: false,
      build: {
        // Exclude <root>/index.html from inputs
        rollupOptions: {},
        emptyOutDir: false,
        lib: {
          name: entry.replace(/-/g, "_").toLowerCase(),
          entry: entry,
          formats: ["umd"],
          fileName: () => entryFilenameToOutput(entry),
        },
      },
    };
    addConfig(mergeConfig(newConfig, pluginOptions.libModeViteConfig ?? {}));
  };

  const { additionalHtmlInputs, additionalScriptInputs } =
    separateAdditionalInputs(pluginOptions.additionalInputs);

  // HTML Pages - multi-page-mode
  const htmlEntries = createEntryList([
    manifest.action?.default_popup,
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
    manifest.browser_action?.default_popup,
    manifest.page_action?.default_popup,
    manifest.sidebar_action?.default_panel,
    additionalHtmlInputs,
  ]);
  if (htmlEntries.length > 0)
    addConfig({
      build: {
        rollupOptions: {
          input: entriesToInputs(htmlEntries),
        },
      },
    });

  // Sandbox - multi-page-mode
  const sandboxEntries = createEntryList([manifest.sandbox?.pages]);
  if (sandboxEntries.length > 0)
    addConfig({
      build: {
        rollupOptions: {
          input: entriesToInputs(sandboxEntries),
        },
      },
    });

  // Every Background Script gets it's own config
  const backgroundScripts = createEntryList([
    manifest.background?.service_worker,
    manifest.background?.page,
    manifest.background?.scripts,
  ]);
  backgroundScripts?.forEach(addLibModeConfig);

  // Every Content Script gets it's own config
  manifest.content_scripts?.forEach((cs: Manifest.ContentScript) => {
    const csEntries = createEntryList([cs.js, cs.css]);
    csEntries.forEach(addLibModeConfig);
  });

  // Each additional script gets it's own config
  additionalScriptInputs?.forEach(addLibModeConfig);

  return configs;
}

function createEntryList(
  a: Array<string | string[] | undefined> | undefined
): string[] {
  return compact<string>(a?.flat() ?? []).filter(
    (entry) => !entry.startsWith(GENERATED_PREFIX)
  );
}

function entriesToInputs(entries: string[]): InputOption {
  return entries.reduce<Record<string, string>>((input, entry) => {
    if (!entry.startsWith(GENERATED_PREFIX))
      input[entry] = entryFilenameToOutput(entry);
    return input;
  }, {});
}

function separateAdditionalInputs(additionalInputs?: string[]) {
  const additionalScriptInputs: string[] = [];
  const additionalHtmlInputs: string[] = [];
  additionalInputs?.forEach((additionalInput) => {
    if (entryFilenameToOutput(additionalInput).endsWith("html"))
      additionalHtmlInputs.push(additionalInput);
    else additionalScriptInputs.push(additionalInput);
  });
  return { additionalScriptInputs, additionalHtmlInputs };
}
