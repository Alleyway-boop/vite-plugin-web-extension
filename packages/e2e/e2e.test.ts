import { testBuild } from "./utils/test-build";
import fs from "fs/promises";
import { UserConfig } from "vite";
import path from "path";
import { expectManifest } from "./utils/expect-manifest";
import { describe, it, expect, beforeEach } from "vitest";
import webExtension from "vite-plugin-web-extension";

const DIST_DIRECTORY = path.resolve(process.cwd(), "dist");

async function expectBuildToMatchSnapshot(
  config: UserConfig,
  expectedDirStructure: string[]
) {
  const res = await testBuild(config);
  for (const file of expectedDirStructure) {
    expect(await fs.lstat(file)).toBeDefined();
  }
  expect(res).toMatchSnapshot();
}

function manifest(overrides: any) {
  return () => ({
    name: "test",
    version: "1.0.0",
    description: "test manifest",
    manifest_version: 2,
    icons: {
      "16": "16.png",
      "48": "48.png",
      "128": "128.png",
    },
    ...overrides,
  });
}
function baseConfig(manifestOverrides: any): UserConfig {
  return {
    root: "extension",
    build: {
      outDir: DIST_DIRECTORY,
      emptyOutDir: true,
    },
    plugins: [
      webExtension({
        manifest: manifest(manifestOverrides),
      }),
    ],
  };
}
function baseOutputs(additionalOutputs?: string[]): string[] {
  return [
    "dist/manifest.json",
    "dist/16.png",
    "dist/48.png",
    "dist/128.png",
    ...(additionalOutputs ?? []),
  ];
}

describe("Vite Plugin Web Extension", () => {
  beforeEach(async () => {
    try {
      await fs.rm(DIST_DIRECTORY, { recursive: true });
    } catch (err) {
      // We just want to delete the directory if it exists, so we don't care about this failing (ie: the dir doesn't exist)
    }
  });

  it("should build a simple popup extension", () =>
    expectBuildToMatchSnapshot(
      baseConfig({
        browser_action: {
          default_popup: "page1.html",
        },
      }),
      baseOutputs(["dist/page1.html"])
    ));

  it("should build a simple background script extension", () =>
    expectBuildToMatchSnapshot(
      baseConfig({
        background: {
          scripts: ["script1.js", "script2.ts"],
        },
      }),
      baseOutputs(["dist/script1.js", "dist/script2.js"])
    ));

  it("should build a extension with both html pages and scripts", () =>
    expectBuildToMatchSnapshot(
      baseConfig({
        browser_action: {
          default_popup: "page1.html",
        },
        background: {
          scripts: ["script1.js", "script2.ts"],
        },
      }),
      baseOutputs(["dist/script1.js", "dist/script2.js"])
    ));

  it.each<[boolean, boolean]>([
    [true, false],
    [true, true],
    [false, false],
    [false, true],
  ])(
    "should not fail when emptyOutDir=%s and the outDirExists=%s",
    async (emptyOutDir: boolean, distExists: boolean) => {
      if (distExists) {
        await fs.mkdir(DIST_DIRECTORY);
        expect((await fs.lstat(DIST_DIRECTORY)).isDirectory()).toBe(true);
      } else {
        await expect(() => fs.lstat(DIST_DIRECTORY)).rejects.toThrowError(
          "ENOENT: no such file or directory"
        );
      }
      await expectBuildToMatchSnapshot(
        {
          root: "extension",
          build: {
            outDir: DIST_DIRECTORY,
            emptyOutDir,
          },
          plugins: [
            webExtension({
              manifest: manifest({
                browser_action: {
                  default_popup: "page1.html",
                },
              }),
            }),
          ],
        },
        baseOutputs(["dist/page1.html"])
      );
    }
  );

  it("should work when the vite root is not specified", () =>
    expectBuildToMatchSnapshot(
      {
        publicDir: "extension/public",
        build: {
          outDir: DIST_DIRECTORY,
          emptyOutDir: true,
        },
        plugins: [
          webExtension({
            manifest: manifest({
              browser_action: {
                default_popup: "extension/page1.html",
              },
            }),
          }),
        ],
      },
      [
        "dist/16.png",
        "dist/48.png",
        "dist/128.png",
        "dist/extension/page1.html",
        "dist/manifest.json",
      ]
    ));

  it("should build additional inputs along side the rest of the manifest", () =>
    expectBuildToMatchSnapshot(
      {
        root: "extension",
        build: {
          outDir: DIST_DIRECTORY,
          emptyOutDir: true,
        },
        plugins: [
          webExtension({
            manifest: manifest({
              browser_action: {
                default_popup: "page1.html",
              },
              background: {
                scripts: ["script1.js"],
              },
            }),
            additionalInputs: ["script2.ts"],
          }),
        ],
      },
      baseOutputs(["dist/page1.html", "dist/script1.js", "dist/script2.js"])
    ));

  it.only.each([
    [
      undefined,
      expect.objectContaining({
        manifest_version: 3,
        action: {
          default_popup: "page1.html",
        },
        content_scripts: [
          {
            matches: ["https://google.com/*"],
            js: ["script1.js"],
          },
        ],
      }),
    ],
    [
      "chrome",
      expect.objectContaining({
        manifest_version: 3,
        action: {
          default_popup: "page1.html",
        },
        content_scripts: [
          {
            matches: ["https://google.com/*"],
            js: ["script1.js"],
          },
        ],
      }),
    ],
    [
      "firefox",
      expect.objectContaining({
        manifest_version: 2,
        browser_action: {
          default_popup: "page1.html",
        },
        content_scripts: [
          {
            matches: ["https://duckduckgo.com/*"],
            js: ["script1.js"],
          },
        ],
      }),
    ],
  ])(
    "should respect the browser flags, removing fields other than %s (defaults to chrome)",
    async (browser, expectedManifest) => {
      await expectBuildToMatchSnapshot(
        {
          root: "extension",
          build: {
            outDir: DIST_DIRECTORY,
          },
          plugins: [
            webExtension({
              manifest: manifest({
                "{{chrome}}.manifest_version": 3,
                "{{firefox}}.manifest_version": 2,
                "{{firefox}}.browser_action": {
                  default_popup: "page1.html",
                },
                "{{chrome}}.action": {
                  default_popup: "page1.html",
                },
                content_scripts: [
                  {
                    matches: [
                      "{{chrome}}.https://google.com/*",
                      "{{firefox}}.https://duckduckgo.com/*",
                    ],
                    js: ["script1.js", "{{other}}.script2.ts"],
                  },
                ],
              }),
              browser,
            }),
          ],
        },
        [] // Don't worry about validating the actual files, other tests handle that
      );
      expectManifest(expectedManifest);
    }
  );

  it("should support supplementing the script build config via scriptViteConfig", () =>
    expectBuildToMatchSnapshot(
      {
        root: "extension",
        build: {
          outDir: DIST_DIRECTORY,
          emptyOutDir: true,
        },
        plugins: [
          webExtension({
            manifest: manifest({
              browser_action: {
                default_popup: "page1.html",
              },
              background: {
                service_worker: "dynamic-import.ts",
              },
            }),
            scriptViteConfig: {
              build: {
                rollupOptions: {
                  output: {
                    inlineDynamicImports: true,
                  },
                },
              },
            },
          }),
        ],
      },
      baseOutputs()
    ));
});
