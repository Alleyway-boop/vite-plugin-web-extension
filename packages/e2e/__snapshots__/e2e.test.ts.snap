// Vitest Snapshot v1

exports[`Vite Plugin Web Extension > Configuration > should build additional inputs along side the rest of the manifest 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually
  2. Building script1.js indvidually
  3. Building script2.ts indvidually

Building page1.html (1/3)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

Building script1.js (2/3)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script1\\"
rendering chunks...
../dist/script1.js   0.00 KiB / gzip: 0.02 KiB

Building script2.ts (3/3)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script2\\"
rendering chunks...
../dist/script2.js   0.00 KiB / gzip: 0.02 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.22 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should not fail when emptyOutDir=false and the outDirExists=false 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually

Building page1.html (1/1)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.18 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should not fail when emptyOutDir=false and the outDirExists=true 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually

Building page1.html (1/1)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.18 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should not fail when emptyOutDir=true and the outDirExists=false 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually

Building page1.html (1/1)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.18 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should not fail when emptyOutDir=true and the outDirExists=true 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually

Building page1.html (1/1)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.18 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should respect the browser flags, removing fields other than chrome (defaults to chrome) 1`] = `
"Building for browser: chrome
vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually
  2. Building script1.js indvidually

Building page1.html (1/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

Building script1.js (2/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script1\\"
rendering chunks...
../dist/script1.js   0.00 KiB / gzip: 0.02 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.25 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should respect the browser flags, removing fields other than firefox (defaults to chrome) 1`] = `
"Building for browser: firefox
vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually
  2. Building script1.js indvidually

Building page1.html (1/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

Building script1.js (2/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script1\\"
rendering chunks...
../dist/script1.js   0.00 KiB / gzip: 0.02 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.26 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should respect the browser flags, removing fields other than undefined (defaults to chrome) 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually
  2. Building script1.js indvidually

Building page1.html (1/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

Building script1.js (2/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script1\\"
rendering chunks...
../dist/script1.js   0.00 KiB / gzip: 0.02 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.25 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should support supplementing the script build config via scriptViteConfig 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually
  2. Building dynamic-import.ts indvidually

Building page1.html (1/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

Building dynamic-import.ts (2/2)
vite v3.1.2 building for production...
transforming...
✓ 3 modules transformed.
rendering chunks...
../dist/dynamic-import.js   0.77 KiB / gzip: 0.50 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.23 KiB"
`;

exports[`Vite Plugin Web Extension > Configuration > should work when the vite root is not specified 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building extension/page1.html indvidually

Building extension/page1.html (1/1)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"extension/page1\\"
rendering chunks...
dist/extension/page1.html   0.26 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
dist/manifest.json   0.19 KiB"
`;

exports[`Vite Plugin Web Extension > Manifest inputs > should build a extension with both html pages and scripts 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually
  2. Building script1.js indvidually
  3. Building script2.ts indvidually

Building page1.html (1/3)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

Building script1.js (2/3)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script1\\"
rendering chunks...
../dist/script1.js   0.00 KiB / gzip: 0.02 KiB

Building script2.ts (3/3)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script2\\"
rendering chunks...
../dist/script2.js   0.00 KiB / gzip: 0.02 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.23 KiB"
`;

exports[`Vite Plugin Web Extension > Manifest inputs > should build a simple background script extension 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building script1.js indvidually
  2. Building script2.ts indvidually

Building script1.js (1/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script1\\"
rendering chunks...
../dist/script1.js   0.00 KiB / gzip: 0.02 KiB

Building script2.ts (2/2)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"script2\\"
rendering chunks...
../dist/script2.js   0.00 KiB / gzip: 0.02 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.19 KiB"
`;

exports[`Vite Plugin Web Extension > Manifest inputs > should build a simple popup extension 1`] = `
"vite v3.1.2 building for production...

Build Steps
  1. Building page1.html indvidually

Building page1.html (1/1)
vite v3.1.2 building for production...
transforming...
✓ 1 modules transformed.
Generated an empty chunk: \\"page1\\"
rendering chunks...
../dist/page1.html   0.26 KiB

✓ All steps completed.

transforming...
✓ 1 modules transformed.
rendering chunks...
../dist/manifest.json   0.18 KiB"
`;
